### PR TITLE
Convert `pthread_t` to C++11 `std::thread`

### PR DESCRIPTION
**Pull request by gaul**
This has better cross-platform support and stronger type-safety.

### CODE REVIEW

**Comment by gaul** [comment]
@siketyan Are you still interested in Windows support?  I wonder if the C++11 modernizations might allow more native Windows instead of using MSYS2?

**Comment by siketyan** [comment]
@gaul Unfortunately I am no longer a Windows user 😢 so I cannot help well about porting for Windows. Thank you for mentioning me anyway.

File: src/curl_multi.cpp [hunk 1]
```diff
@@ -21,6 +21,8 @@
 #include <cstdio>
 #include <cstdlib>
 #include <cerrno>
+#include <future>
+#include <thread>
```
Comment (ggtakec) [inline src/curl_multi.cpp:25]: These two header files are defined in `curl_multi.h`, so there is no need to include them here.

File: src/openssl_auth.cpp [hunk 2]
```diff
@@ -26,7 +26,6 @@
 #include <cstdlib>
 #include <cerrno>
 #include <mutex>
-#include <pthread.h>
 #include <unistd.h>
```
Comment (ggtakec) [inline src/openssl_auth.cpp:29]: If you want to make the corrections you commented in subsequent reviews, you will need to `#include <thread>` here.

File: src/openssl_auth.cpp [hunk 3]
```diff
@@ -96,14 +95,6 @@ static void s3fs_crypt_mutex_lock(int mode, int pos, const char* file, int line)
...
 }
 
-static unsigned long s3fs_crypt_get_threadid() __attribute__ ((unused));
-static unsigned long s3fs_crypt_get_threadid()
-{
-    // For FreeBSD etc, some system's pthread_t is structure pointer.
-    // Then we use cast like C style(not C++) instead of ifdef.
-    return (unsigned long)(pthread_self());
-}
-
```
Comment (ggtakec) [inline src/openssl_auth.cpp:106]: This function is necessary.
However, it requires changes to std::thread as follows:
```
static unsigned long s3fs_crypt_get_threadid() __attribute__ ((unused));
static unsigned long s3fs_crypt_get_threadid()
{
    std::hash<std::thread::id> hash;
    return (unsigned long)hash(std::this_thread::get_id());
}
```

File: src/openssl_auth.cpp [hunk 4]
```diff
@@ -144,7 +135,6 @@ bool s3fs_init_crypt_mutex()
     s3fs_crypt_mutex = new std::mutex[CRYPTO_num_locks()];
     // static lock
     CRYPTO_set_locking_callback(s3fs_crypt_mutex_lock);
-    CRYPTO_set_id_callback(s3fs_crypt_get_threadid);
```
Comment (ggtakec) [inline src/openssl_auth.cpp:147]: The call to this function must remain.

File: src/openssl_auth.cpp [hunk 5]
```diff
@@ -162,7 +152,6 @@ bool s3fs_destroy_crypt_mutex()
     CRYPTO_set_dynlock_destroy_callback(nullptr);
     CRYPTO_set_dynlock_lock_callback(nullptr);
     CRYPTO_set_dynlock_create_callback(nullptr);
-    CRYPTO_set_id_callback(nullptr);
```
Comment (ggtakec) [inline src/openssl_auth.cpp:165]: The call to this function must remain.

**Review by ggtakec — CHANGES_REQUESTED** [review]
I would like you to consider fixing `s3fs_crypt_get_threadid`.
This call should be required for OpenSSL's multithread support.
(However, I checked and it does not affect the error that currently occurs in `sanitize_thread`. = The error continues to occur, so we are investigating it separately.)

File: src/openssl_auth.cpp (hunk 3 above)
Comment (gaul) [reply src/openssl_auth.cpp:106]: I'm not sure this is actually used in OpenSSL 1.0 and later: https://linux.die.net/man/3/crypto_set_id_callback but I added it as you suggested.

File: src/curl_multi.cpp (hunk 1 above)
Comment (ggtakec) [reply src/curl_multi.cpp:25]: OK, that's fine.

File: src/openssl_auth.cpp (hunk 3 above)
Comment (ggtakec) [reply src/openssl_auth.cpp:106]: Thank you for adding it.
I'm also concerned, but there's the datarace issue that's also mentioned in other issues, so I'd like to keep it.

**Pull request merged**